docs/geps/gep-04.rst

```diff
@@ -0,0 +1,177 @@
+   more global like replacing all transfers with a universal basic income.
+
+2. Computing taxes and transfers with GETTSIM takes a long time. The major reason is
+   that the implementation in version 0.2 does not use vectorization. The second reason
+   is that it is not possible to limit the computations to a set of target variables,
+   which the researcher is ultimately interested in.
+
+
+Solution
+--------
+
+At the core of the solution is the following observation. To compute variables in the
+tax and transfer system, the user starts with some input data. A variable in the tax and
+transfer system can be computed by a function which receives the data and parameters as
+inputs and yields the variable. Since a single function might become too complex, we
+could split up the function into multiple functions which receive the data and
+parameters and compute intermediate outcomes. Then, the task is to call the collection
+of functions in the correct order and pass in the correct data and parameters. The last
+function in this collection will receive the intermediate outcomes and calculate the
+requested target variable.
+
+Splitting complex calculations into smaller pieces has a lot of the usual advantages of
+why we use functions in the first place: readability, simplicity, lower maintenance
+costs (see single-responsibility principle). Another advantage is that each function is
+a potential entry point for a researcher to change the tax and transfer system if she is
+able to replace this function with her own version.
+
+Two questions arise.
+
+1. How do we know which data and parameters to pass to the functions?
+
+To solve the first problem, we were inspired by `pytest and its fixtures
+<https://docs.pytest.org/en/stable/fixture.html>`_. The idea is that every variable has
+a name which is also a valid variable name in Python (cannot start with number,
+alphanumeric, lower- and uppercase, underscores). Then, a function inside the tax and
+transfer system uses this name as an input argument. See the following, intentionally
+abstract and incorrect example.
+
+.. code-block:: python
+
+    def kindergeld(anz_kinder_hh, kindergeld_params):
+        pass
+
+The function :func:`kindergeld` requires the variable ``anz_kinder_hh`` which is the
+number of children per household. When the function is called, a :class:`pandas.Series`
+with the information is passed to this argument. The same happens to the parameters
+where ``"kindergeld"`` is a single parameter-topic inside the collection of all
+parameters and ``kindergeld_params`` is a dictionary.
+
+The result of this function is again a :class:`pandas.Series` which has the name
+``kindergeld``, the same name as the function. Another function, say
+
+.. code-block:: python
+
+    def benefits(kindergeld, arbeitsl_geld_2, params):
+        pass
+
+would need to have ``kindergeld`` as a name for an input argument to request this
+:class:`pandas.Series`.
```

Comment:
Make sure somewhere here, that valid inputs are: parameter values, input variables, other functions, summations of other functions (with the new change proposed in #281)